src/devices/machine/session_time.cpp

```diff
@@ -0,0 +1,182 @@
+// license:BSD-3-Clause
```

Comment:
I understand the need for something like this and I like the feature set, but I'm not sure this is exactly the right way to do it.  I need architecture advise from @cuavas on this.  It feels like something that should be in the core rather than a device, and it shouldn't assume that its users are printers (I could see it potentially being used for plotters, paper tape and card punches, and stuff of that nature).

Reply:
I see there's already similar functionality to convert tags to filenames (looking at my nvram directories) and I wanted a timestamp that would allow filename sorting to keep things in order.  So when I launch the file viewer from Ubuntu's file manager (which I think is eog ) when I hit the right arrow I get the next image in the sequence.  If I type "eog snap/ap2000" from the command line, and hit the left arrow I get the most recent image printed.

nvram/apple2e:
sl1_grappler_prn_ap2000_eeprom
sl1_grappler_prn_etibuffer_prn_ap2000_eeprom
sl1_grappler_prn_etibuffer_prn_etibuffer_prn_ap2000_eeprom
sl1_grapplus_prn_ap2000_eeprom



I also notice that there's a timestamp functionality in video.cpp's open_next

//-------------------------------------------------
//  open_next - open the next non-existing file of
//  type filetype according to our numbering
//  scheme
//-------------------------------------------------

std::error_condition video_manager::open_next(emu_file &file, const char *extension, uint32_t added_index)
{
...
        // handle %t in the template (for timestamp)
        std::string snaptime("%t");
        int pos_time = snapstr.find(snaptime);

        if (pos_time != -1)
        {
                char t_str[15];
                const std::time_t cur_time = std::time(nullptr);
                strftime(t_str, sizeof(t_str), "%Y%m%d_%H%M%S", std::localtime(&cur_time));
                strreplace(snapstr, "%t", t_str);
        }

Reply:
Maybe bitmap_printer_filename.cpp?

Reply:
My string functions to change colons to underscores aren't really pretty, so I should replace them with something from 

std::string running_machine::nvram_filename(device_t &device) const

               std::string tag(device.tag());
                tag.erase(0, 1);
                strreplacechr(tag,':', '_');

Reply:
I was thinking about making a copy of video_manager::open_next and adding some new % specifiers like  %f = full tag  %b = basename  %s=session time  %p = page number along with a default snapname.

In the meantime, I made a version of write_snapshot_to_file that would just call video_manager::open_next and save it in the driver's directory like it was a regular F12 snapshot.

void bitmap_printer_device::write_snapshot_to_file(std::string directory, std::string name)
{
	emu_file file(machine().options().snapshot_directory(), OPEN_FLAG_WRITE | OPEN_FLAG_CREATE | OPEN_FLAG_CREATE_PATHS);
	std::error_condition const filerr = machine().video().open_next(file, "png");

	if (!filerr)
	{
		static const rgb_t png_palette[] = { rgb_t::white(), rgb_t::black() };

		// save the paper into a png
		util::png_write_bitmap(file, nullptr, m_page_bitmap, 2, png_palette);
	}
}